### ISSUE TITLE
Using `new TrackedArray<string>()` in TypeScript

### ISSUE THREAD

**Issue by richard-viney** [tracked-tools/tracked-built-ins#23]
Hi,

Is it expected to be able to write `new TrackedArray<string>()` in TypeScript?

I'm seeing the following:

```typescript
// This works fine. The type of foo is string[].
const foo = TrackedArray.of<string>()
foo.push("item")

// This gives the error: "Property 'push' does not exist on type 'TrackedArray<string>'. ts(2339)".
// The type of bar is `TrackedArray<string>`
const bar = new TrackedArray<string>()
bar.push("item")
```

Should `TrackedArray<T>` be extending `Array<T>` in the type definitions?

Thanks.

**Comment by pzuraq** [comment]
hmm, I'm not sure it should, because we don't want to deopt the array constructor unintentionally by forcing an unrelated object to use its constructor (it'll have a different shape, so it could cause the array constructor, which is pre-JITed, to deopt when it sees its not a standard array).

It may be that we need to define `TrackedArray` as an interface and use some tricks to make TS think the class matches the interface. @dfreeman had a similar suggestion for `TrackedObject`: [playground](https://www.typescriptlang.org/play/index.html?target=99#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzDigxAAUYcAPATwB4AVeESk1YAZ3hwCMArcDAD4AFD14AuePQCUk+gG4AsACgVWVCRiIoYBPRg6A1iGAB5PgIDCedhhjJMOGPADeK+B-ioQAd3gMmFhA2ThcAXxExOVkpJWVPJAoAWwBRDRgsEHZaAGlA1g54WwzUAHN4AB8i6iTuHAhKr2RakBgAGngANRFgu0z2SQBJTShuCBBaImA8CGp4AG0cjs6AXUEYgCVwJ2Bc5cE4sJUVFHRsPCkDMGMzC0wA5gLOMQFIvmi5V3dPcYx4AAcKBgcPAALzwcz8TAAOlKIAw5BwwIw1H+IFMiFEfGkcW+HgA9Pj4AABDDsAC0WFKqCcIEkYCgqAA5H92PD4EyIFhuEz4Op4BgABYIf7QailChoYAdQbwUrILLsPHwX7wUDsMCccGQgSw+GmHyoRFomAogAiWUIWH+wJg7CxvBxx3iP3ZkDwCG1dww0MIIGIIGEgKROCdqhdHkQTngwlVwf+fPw6s10i+Ec8OphoEQ6jIFBNKOE7u8HXjHWT7Hm8ZWYYSR3DCTgGGQMHwfoDiJoRYgHrD9ZU+iMJkzPsQyTSfSyYKQaEwuHwY5wqXS-Vy+WChWK6nKVXYNTqDSqqGa3Fa+2EvQyWSGIzGEymMzmi2Wa022xguyWXUEaYShJJZKUtStKSFApRQPyQoIFy3ACoK27KqqYjTiO0KLsuk72pe-S1p4TYtm2RAkJ21AOn2zrMP8Th-NmUDIBAfyDtcw7evAUCcGghg0oabGcExNwjtYqDFA4tryEAA)

**Comment by richard-viney** [comment]
The following seems to work for me:

```ts
declare module "tracked-built-ins" {
  interface TrackedArray<T> extends Array<T> {}

  export const TrackedArray: new <T>(array?: T[]) => TrackedArray<T>
}
```

However, I'm not certain how to get `from()` and `of()` in there as well. For my use case this doesn't matter though.

If I use the extended `tracked` wrapper in this package and do `const myArray = tracked([])` then I have the same typing problems because that function returns a `TrackedArray<T>` which has essentially no methods on it. If it returned `T[]` I think it would work, should the return types [here](https://github.com/pzuraq/tracked-built-ins/blob/master/addon/-private/decorator.ts#L13) in `decorator.ts` be changed?

**Comment by richard-viney** [comment]
Actually the following is sufficient (I'm currently only using `TrackedArray` from this package):

```ts
declare module "tracked-built-ins" {
  export class TrackedArray<T> extends Array<T> {}
}
```

I understand that the library doesn't want to do `class TrackedArray extends Array {...}` internally for performance reasons, but as far as typings go the above seems potentially simplest?

**Comment by richard-viney** [comment]
Ended up adding the following to prevent a subtle error I encountered while using this library. Any thoughts on this?

```ts
declare module "tracked-built-ins" {
  export class TrackedArray<T> extends Array<T> {
    /**
     * This member protects against implicitly converting a variable of type
     * `Array<T>` to a variable of type `TrackedArray<T>` which would silently
     * lose autotracking. It prevents errors like the following:
     *
     *   class Foo {
     *     private names = new TrackedArray<string>();
     *
     *     setNames(newNames: string[]) {
     *       this.names = newNames;
     *     }
     *   }
     */
    private readonly _trackedBuiltIn: never
  }
}
```

**Comment by chriskrycho** [comment]
I *believe* this was resolved by the changes made in #56 and #161 and released in [2.0](https://github.com/tracked-tools/tracked-built-ins/releases/tag/v2.0.0). Can you confirm that that resolves the issues you were having, @richard-viney?

**Comment by richard-viney** [comment]
Yep all seems to be sorted in 2.0. Thanks!
